### ISSUE TITLE
ALX: a function that copies a string (_strncpy)

### ISSUE THREAD

**Issue by ahmed149** [ahmed149/learning-programing-in-C#32]
#include "main.h"

/**
 * _strncpy - Copies at most an inputted number
 *            of bytes from string src into dest.
 * @dest: The buffer storing the string copy.
 * @src: The source string.
 * @n: The maximum number of bytes to copied from src.
 *
 * Return: A pointer to the resulting string dest.
 */
char *_strncpy(char *dest, char *src, int n)
{
	int i = 0, src_len = 0;

	while (src[i++])
		src_len++;

	for (i = 0; src[i] && i < n; i++)
		dest[i] = src[i];

	for (i = src_len; i < n; i++)
		dest[i] = '\0';

	return (dest);
}

**Comment by ahmed149** [comment]
Function Purpose:

_strncpy is used for copying characters from one string (src) to another string (dest).
It has an additional parameter n, which specifies the maximum number of characters to copy from src to dest.
Function Signature:

char *_strncpy(char *dest, char *src, int n): This line is like a title for the function. It takes three parameters: dest (the destination where we want to copy), src (the source from which we want to copy), and n (the maximum number of characters to copy).
Variable Explanation:

i is a counter variable used to iterate through the characters in the src and dest strings.
src_len is used to store the length of the src string.
Finding the Length of src:

The code starts by finding the length of the src string. It does this by using a while loop that iterates until it reaches the null terminator ('\0') in src. This step is necessary to know how many characters are there in src.
Copying Characters from src to dest:

After finding the length of src, the function enters another loop. This loop does two things:
It copies characters from src to dest up to a maximum of n characters (controlled by the condition i < n).
If it encounters the null terminator in src before n characters are copied, it stops copying.
Null-Terminating dest:

After copying characters from src to dest, the function ensures that dest is null-terminated by adding null characters ('\0') to the end of dest until the total length of dest is n. This is done in the final for loop.
Return Value:

Finally, the function returns a pointer to the modified dest string.
Example:
Let's say:

dest is a character array with enough space to hold the result.
src is "Hello, world!".
n is 5.
After using _strncpy(dest, src, n), dest will contain "Hello", which is the first 5 characters of src. The rest of dest is filled with null characters to ensure it's null-terminated.

In simpler terms, _strncpy is like making a copy of the first n characters from one string (src) into another string (dest).

**Comment by ahmed149** [comment]
#include <stdio.h>

/**
 * _strncpy_recursive - Recursively copy up to 'n' characters from src to dest.
 * @dest: A pointer to destination input.
 * @src: A pointer to source input.
 * @n: The maximum number of characters to copy from src to dest.
 *
 * Return: A pointer to the resulting string dest.
 */

char *_strncpy_recursive(char *dest, char *src, int n)
{
    // If n is zero or we've reached the end of src, stop the recursion.
    if (n == 0 || *src == '\0')
    {
        *dest = '\0'; // Null-terminate dest.
    }
    else
    {
        *dest = *src; // Copy the current character from src to dest.
        _strncpy_recursive(dest + 1, src + 1, n - 1); // Recurse with the next characters and decrement n.
    }

    return dest;
}

int main()
{
    char dest[50]; // Initialize dest with enough space.
    char src[] = "Hello, world!";
    int n = 5; // Example: Copy the first 5 characters from src.

    printf("Original dest: %s\n", dest);
    _strncpy_recursive(dest, src, n);
    printf("Copied dest: %s\n", dest);

    return 0;
}

**Comment by ahmed149** [comment]
In this recursive version:

We have a parameter n that specifies the maximum number of characters to copy from src to dest.

We check two conditions for stopping the recursion:

If n becomes zero (meaning we've copied the desired number of characters), or
If we've reached the end of src (null terminator).
If neither of these conditions is met, we copy the current character from src to dest and then recursively call _strncpy_recursive with the next characters (dest + 1, src + 1) and decrement n.

The result is that it copies up to n characters from src to dest.

Here's how it works in practice:

If dest is an empty character array, src is "Hello, world!", and you specify n as 5, the result will be "Hello". It copies the first 5 characters from src to dest.
